### PR TITLE
Optimise leader pod check

### PR DESCRIPTION
**Pull request by gytis**
@iocanel @steven-sheehy

### CODE REVIEW

File: spring-cloud-kubernetes-leader/src/main/java/org/springframework/cloud/kubernetes/leader/LeaderProperties.java [hunk 1]
```diff
@@ -61,12 +58,6 @@
 	 */
 	private String configMapName = DEFAULT_CONFIG_MAP_NAME;
 
-	/**
-	 * Kubernetes labels common to all leadership candidates.
-	 * Default: empty
-	 */
-	private Map<String, String> labels = Collections.emptyMap();
-
```
Comment (steven-sheehy) [inline spring-cloud-kubernetes-leader/src/main/java/org/springframework/cloud/kubernetes/leader/LeaderProperties.java]: Can you add enabled property that is checked in LeaderAutoConfiguration?

Comment (gytis) [reply spring-cloud-kubernetes-leader/src/main/java/org/springframework/cloud/kubernetes/leader/LeaderProperties.java]: It is already here https://github.com/spring-cloud-incubator/spring-cloud-kubernetes/blob/master/spring-cloud-kubernetes-leader/src/main/java/org/springframework/cloud/kubernetes/leader/LeaderProperties.java#L46

Comment (steven-sheehy) [reply spring-cloud-kubernetes-leader/src/main/java/org/springframework/cloud/kubernetes/leader/LeaderProperties.java]: You're referring to autoStartup. I'm referring to "spring.cloud.kubernetes.leader.enabled", which is missing from LeaderProperties but checked in this line `@ConditionalOnProperty(value = "spring.cloud.kubernetes.leader.enabled", matchIfMissing = true)`

Comment (gytis) [reply spring-cloud-kubernetes-leader/src/main/java/org/springframework/cloud/kubernetes/leader/LeaderProperties.java]: Done. Although, it's not strictly necessary.

**Comment by gytis** [comment]
Unrelated Ribbon tests have failed.

**Comment by gytis** [comment]
Retest this please

**Comment by steven-sheehy** [comment]
I know above is for circleci, but I also tested this change and think it may pose problems. If I start two pods at the exact same time then both will transition to leader at the same time. The problem corrects itself at the next poll period, but we should never have more than one leader at a time. I think this occurs due to the ready check since both pods are not ready so they both acquire.

If we want to keep ready check, then logic should ensure leader.isValid() returns true if he is ready OR **if he is not ready but there are no other ready pods**.

Also, it was pretty difficult to diagnose this since there is very little logging. Anyway some more debug statements could be added like when attempting acquire/revoke, updating configmap, etc.?

**Comment by gytis** [comment]
Were you able to reproduce the issue with two leaders? Did you already have a configmap created?

**Comment by steven-sheehy** [comment]
Yes, it occurs consistently as long as your pod's readinessProbe.initialDelaySeconds is high enough. The configmap was already created at the time.

**Comment by steven-sheehy** [comment]
1) Just to expand upon the problem of checking only pod exists: If you have two ready pods and one restarts, it will continue to keep the same pod name and most likely keep the leader lock. But that pod is not really able to respond to requests yet and it would be much better to elect the pod that is currently up and ready instead.

2) "Only allow pod to become a leader if it is ready." This is what your PR currently does. I think it needs to additionally check if not ready, then proceed if no other ready pods.

Maybe we can sidestep the issue and instead of checking ready status (which is mostly applicable to service traffic), it would be better to look at pod.status.phase == 'Running'. Main thing is to ensure pod not elected leader if he's still starting up, not that he became unready after startup. I don't think pod phase can ever go back to another status without a restart. Kubernetes docs say that Running means:

> The pod has been bound to a node, and all of the containers have been created. At least one container is still running, or is in the process of starting or restarting.

**Comment by steven-sheehy** [comment]
Nevermind about the pod.status.phase. That is marked as running even when not ready. So guess we still need to figure out how to use ready.

**Comment by iocanel** [comment]
retest this please

**Comment by iocanel** [comment]
IMHO, we should check if a pod is ready. The client has a utility called `Readiness` that can easily perform the check.

The question is what happens when a pod that was once ready, fails its readiness check....
I think that all candidates should watch for pod events and if the leader is no longer ready, should automatically kick an election.

**Comment by geoand** [comment]
Should the PR be rebased onto the latest master (or have master merged into it) in order to take advantage of the fix to the Ribbon test?

**Comment by iocanel** [comment]
@gytis can u please rebase?

**Comment by gytis** [comment]
OK so I'll have to reimplement the logic how leadership is given and revoked. Will also rebased at the same time.

**Pull request closed without merging**